x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -727,9 +727,9 @@ private void asyncNodeOperation(NodeTermsEnumRequest request, Task task, ActionL
                         false
                     )
                 );
-            }
-            if (canMatchShard(shardId, request) == false) {
-                // Permission denied or can't match, remove shardID from request
+                request.remove(shardId);
+            } else if (canMatchShard(shardId, request) == false) {
+                // if can't match, remove shardID from request as well
```

Comment:
It does seem weird to me that we didn't remove this previously. Do we know why? I guess it was just a bug?

Reply:
I introduced this in this enhancement PR https://github.com/elastic/elasticsearch/pull/91720/files#diff-e4fdb92642f37bf8c1336593974e437acdb8da3a7e05f65fcd1efb83847eb92bR733 because I separated the two previous conditions into two branches but forgot to do the remove in both branches then.